### PR TITLE
[main-theme] Refactor to be testable for each events

### PR DESCRIPTION
**Pull request by nullist0**
### Motivation

For the testable application. we can do the following things.

- Remove mutable data classes.
- Separate logics from `Service` to each classes.

These leads the logics to be more testable without many dependencies. Also, this approach can be found in the theory.

#### Domain Event from DDD

In DDD, the domain event is defined as followings. In short, the **domain events** are events related to the changes in the domain models(aggregates), which domain experts care about.

> A domain event is a full-fledged part of the domain model, a representation of something that happened in the domain. Ignore irrelevant domain activity while making explicit the events that the domain experts want to track or be notified of, or which are associated with state changes in the other model objects.

The Chord algorithm shows the solution of the P2P hash table. I think the domain model is Chord Node, and the domain event is methods of managing the key-values. So I designed each logics. which are setting values / getting values / finding the nodes, as a domain events.

#### UseCase from Clean Architecture 

Also, there are some inferences from the Clean architecture.

> The software in this layer contains application specific business rules. It encapsulates and implements all of the use cases of the system. These use cases orchestrate the flow of data to and from the entities, and direct those entities to use their enterprise wide business rules to achieve the goals of the use case.

With the same argument, the use case of Chord algorithm is managing key-values. So they can be seen as use cases.

### Modifications

- Address the commits from the previous PR.
  - Add `InvalidChordNode`.
  - Remove the mutable `ChordNode`.
- Separate the domain events from `Service`.
- Add interfaces to handle the local datas.

### Result

We can have separated logics to each classes, and they represent the domain event or use case. Also we can get each test codes without each classes.

### Next works

- Remove too many type arguments if it is not needed.
- Implement Repository to remove values so that the new node has its data.

### Reference

- https://medium.com/@chaojie.xiao/domain-driven-design-practice-domain-events-15b38f3c58fc
- https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html

### CODE REVIEW

File: hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.lia.dht.nullist0.chord.event
+
+import org.lia.dht.nullist0.chord.ChordConfiguration
+
+/**
+ * An interface to represent the domain event of chord algorithm.
+ */
+internal interface ChordEvent<Id, Value, ReturnType>: Function<ReturnType> {
```
Comment (TheWeaVer) [inline hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt:8]: Is there a plan to create `JoinToChordRing`?

File: hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package org.lia.dht.nullist0.chord.event
+
+import org.lia.dht.nullist0.chord.ChordConfiguration
+
+/**
+ * An interface to represent the domain event of chord algorithm.
```
Comment (TheWeaVer) [inline hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt:6]: How about...?

```suggestion
 * An interface to represent the operators of chord algorithm.
```

File: hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/test/kotlin/org/lia/dht/nullist0/chord/event/FindSuccessorByIdTest.kt [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+        // when
+        val successor = findSuccessorById(configuration)
+
+        // then
+        assertEquals(myNode, successor)
+    }
+
+    @Test
+    fun testFindSuccessorById_twoNodes_with_otherNode_dataNode_myNode() = runTest {
```
Comment (TheWeaVer) [inline hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/test/kotlin/org/lia/dht/nullist0/chord/event/FindSuccessorByIdTest.kt:79]: Q. It's inconsistent with upper test method name. Why do you prefer to mixuse camelCase and snake_case ? `testFindSuccessorById_twoNodesWithOtherDataNodeAndMyNode` is my understanding of your intention.

**Review by TheWeaVer — COMMENTED** [review]
Some of test classes require new line end of files, please check it also. 😄

File: hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt (hunk 1 above)
Comment (TheWeaVer) [reply hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt:8]: Thank you for implementing interesting approach 😄

File: hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+ */
+internal interface ChordEvent<Id, Value, ReturnType>: Function<ReturnType> {
+    /**
+     * A function to implement a domain logic for each event.
```
Comment (TheWeaVer) [inline hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt:10]:
```suggestion
     * A function to implement a specific logic for each event.
```

File: hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt (hunk 1 above)
Comment (nullist0) [reply hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt:8]: Yes there is a plan to create `JoinToChordRing` and `LeaveFromChordRing`.

Comment (TheWeaVer) [reply hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/event/ChordEvent.kt:8]: Nice 🙂

**Review by TheWeaVer — APPROVED** [review]
LGTM 👍

**Pull request merged**